### PR TITLE
lis2dw12: add BW filter, low noise, threshold HP_REF and autoreset

### PR DESCRIPTION
**Pull request by maximevince**
Added support for a few of the more advanced features present in the LIS2DW12 accelero:
* Digital filtering bandwidth selection
* Wakeup threshold trigger
* High-Pass reference mode
* High-Pass reference auto-reset on wakeup interrupt

Let me know if the split between DeviceTree config (like the bw_filt value) and the Kconfig defines is okay.

### CODE REVIEW

File: drivers/sensor/lis2dw12/Kconfig
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dw12/Kconfig:68]: What's going on here? This is in Kconfig, not DT.

Comment (avisconti) [reply drivers/sensor/lis2dw12/Kconfig:68]: Yes, everything should be done in DT. I don't see changes in the dts/binding. Am I missing something?

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (avisconti) [inline drivers/sensor/lis2dw12/lis2dw12.c]: Wherte is bw_filt assigned? Should be taken from DT right?

**Review by avisconti — CHANGES_REQUESTED** [review]

**Comment by avisconti** [comment]
@maximevince 
Please also change the commit header format

**Comment by maximevince** [comment]
Thanks for the feedback, I will make sure to process the changes requested. It will take some time, as I am on leave now.

**Comment by maximevince** [comment]
> @maximevince Please also change the commit header format

Excuse my ignorance, but could you give me an example of a correct commit header format?

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (maximevince) [reply drivers/sensor/lis2dw12/lis2dw12.c]: This part was missing, indeed. I have pushed a new commit where this is fixed.

**Comment by avisconti** [comment]
> > @maximevince Please also change the commit header format
> 
> Excuse my ignorance, but could you give me an example of a correct commit header format?
> 
> EDIT: I have guessed it was the `drivers/sensor: lis2dw12:` prefix you were referring to, so I have added those. Let me know if it's okay now.

Please take a look at https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines
Basically, you should have:

```
commit title
<blank line>
commit body (multiple lines)
<blank line>
signature
```

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (avisconti) [reply drivers/sensor/lis2dw12/lis2dw12.c]: OK, and I think you can get rid of CONFIG_LIS2DW12_BW_FILT in Kconfig.
Again, the default will be 0 if not defined in a particular driver instance.

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (avisconti) [inline drivers/sensor/lis2dw12/lis2dw12.c:398]: Again, define it in DT.

**Review by avisconti — COMMENTED** [review]
I think you should also re-organize the commits per feature, like wake-up support, tap support, ...

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (maximevince) [reply drivers/sensor/lis2dw12/lis2dw12.c:366]: That makes sense!

**Comment by maximevince** [comment]
@avisconti : I have restructured the commits to contain one (set of) feature(s) per commit, and re-did the commits messages. Commits are now signed and GPG-signed as well.

I have left one Kconfig option in there: `CONFIG_LIS2DW12_THRESHOLD`, which allows to disable this portion of the code, in case a smaller code size is desired and the threshold feature is not used.

All other config options have been moved to DT.

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dw12/lis2dw12.c:383]: LOG_INF seems a bit much; why did you select this instead of LOG_DBG like the other calls in this same function?

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dw12/lis2dw12.c:393]: Similar comments in this patch -- LOG_INF seems a bit noisy for these, no?

File: drivers/sensor/lis2dw12/lis2dw12.h
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dw12/lis2dw12.h:73]: It's @avisconti 's driver so this is not a blocking request, but it seems like you could save a little flash by bit-packing these.

File: drivers/sensor/lis2dw12/lis2dw12_trigger.c
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dw12/lis2dw12_trigger.c:78]: You get the idea about LOG_INF being probably too noisy -- I'll stop mentioning it

**Review by mbolivar-nordic — DISMISSED** [review]
Devicetree bindings changes are well done. Nice job.

A few random comments but none are blocking. OK for me on the DT side, so I'm approving under the assumption that @avisconti should +1 before merge

**Review by avisconti — DISMISSED** [review]

**Comment by avisconti** [comment]
Very good job indeed!

**Comment by maximevince** [comment]
Okay, one more push where I changed all of the added `LOG_INF`s to `LOG_DBG`s.
If it's okay for you guys, it can be merged!

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (maximevince) [reply drivers/sensor/lis2dw12/lis2dw12.c]: Changed them to LOG_DBG. Totally makes sense!

**Comment by MaureenHelm** [comment]
> Okay, one more push where I changed all of the added `LOG_INF`s to `LOG_DBG`s.
> If it's okay for you guys, it can be merged!

These changes were made in the wrong commit.

There's also a compliance check that needs to be fixed. Everything else looks great

**Comment by maximevince** [comment]
> > Okay, one more push where I changed all of the added `LOG_INF`s to `LOG_DBG`s.
> > If it's okay for you guys, it can be merged!
> 
> These changes were made in the wrong commit.
> 
> There's also a compliance check that needs to be fixed. Everything else looks great

Thanks @MaureenHelm . Issues have been addressed!

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (gmarull) [inline drivers/sensor/lis2dw12/lis2dw12.c:383]: just use `%u`, or, `PRIu8`

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (gmarull) [inline drivers/sensor/lis2dw12/lis2dw12.c:319]: please propagate error codes if you're not handling them,

```
ret = f();
if (ret < 0) {
    return ret;
}
```

applies to others

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (gmarull) [inline drivers/sensor/lis2dw12/lis2dw12.c:211]: magic constants here, please, define or document them.

File: drivers/sensor/lis2dw12/Kconfig
Comment (gmarull) [inline drivers/sensor/lis2dw12/Kconfig:63]: please add some help here

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (gmarull) [inline drivers/sensor/lis2dw12/lis2dw12.c:227]: missing `__fallthrough`

**Review by gmarull — DISMISSED** [review]
some nitpicks

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (maximevince) [reply drivers/sensor/lis2dw12/lis2dw12.c:319]: I just mirrored what the rest of the driver was doing.
I'll refactor my part.

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (maximevince) [reply drivers/sensor/lis2dw12/lis2dw12.c:162]: Not sure I understand the suggestion.
Using `sensor_ms2_to_g` would make me loose a lot of precision; everything after the decimal dot, right?
Or are you suggesting to scale the input to `sensor_ms2_to_g` before calling the conversion?

File: drivers/sensor/lis2dw12/Kconfig
Comment (maximevince) [reply drivers/sensor/lis2dw12/Kconfig:63]: Ok, done!

File: drivers/sensor/lis2dw12/lis2dw12.c
Comment (maximevince) [reply drivers/sensor/lis2dw12/lis2dw12.c:247]: `__fallthrough` for consecutive cases in a switch? (without any other lines of code in between)
Sure, I can do that, but:
- doesn't that hurt readability more, than help it?
- I can't see that being done almost anywhere else in `drivers/sensor`

**Review by MaureenHelm — DISMISSED** [review]

**Comment by maximevince** [comment]
Addressed a few style issues + fixed a bug in setting the threshold interrupt and value.

**Comment by maximevince** [comment]
Anything missing from my side to get this merged? @avisconti @mbolivar-nordic @gmarull @MaureenHelm

**Comment by mbolivar-nordic** [comment]
@gmarull is out on leave, so I will dismiss his review if @MaureenHelm or a sensor collaborator can give this another +1

**Comment by avisconti** [comment]
> @avisconti : I have restructured the commits to contain one (set of) feature(s) per commit, and re-did the commit messages. Commits are now signed and GPG-signed as well.
> 
> I have left one Kconfig option in there: `CONFIG_LIS2DW12_THRESHOLD`, which allows to disable this portion of the code when not selected, which results in smaller code size, when the threshold feature is not used.
> 
> All other config options have been moved to DT.

But once enabled it is enabled for all driver instance. I usually have same comment here, which is that there might be the need to have one instance configured in one way and the other not. Isn't it?
Moreover the LIS2DW12 is a plain accelerometer, which means that there may be lot of application using multiple instances.

File: drivers/sensor/lis2dw12/Kconfig
Comment (avisconti) [inline drivers/sensor/lis2dw12/Kconfig:73]: Please reconsider to have also this one in DT, or convince me that is not needed.

**Review by avisconti — CHANGES_REQUESTED** [review]
Everything is good but the THRESHOLD in Kconfig.
Nevertheless, I'm ready to change my mind if this choice is supported by good arguments. In my opinion the flexibility to select different configs for different driver instances is more important than having a smaller code size.

File: drivers/sensor/lis2dw12/Kconfig
Comment (maximevince) [reply drivers/sensor/lis2dw12/Kconfig:73]: Hi @avisconti .
The Kconfig choice here is about the firmware feature: code compiled in or not,
not about enabling/configuring this feature for a specific LIS2DW12 instance.

My board also has two LIS2DW12's on it, so I can still choose to only configure one in threshold mode. The user still needs to call `lis2dw12_attr_set_thresh()` on their specific instance to configure the feature.

It was mainly done from a code-size perspective, hope this is acceptable.

Comment (avisconti) [reply drivers/sensor/lis2dw12/Kconfig:73]: > Hi @avisconti . The Kconfig choice here is about the firmware feature: code compiled in or not, not about enabling/configuring this feature for a specific LIS2DW12 instance.
> 
> My board also has two LIS2DW12's on it, so I can still choose to only configure one in threshold mode. The user still needs to call `lis2dw12_attr_set_thresh()` on their specific instance to configure the feature.
> 
> It was mainly done from a code-size perspective, hope this is acceptable.

Oh yes, sorry. It is called at runtime.
My bad...

**Review by avisconti — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]
@gmarull please revisit

**Pull request merged**
